### PR TITLE
Expand logging capabilities

### LINKED ISSUE #129: Armor stand interactions are not logged
### Expected behavior

After taking armor from an armor stand, this will be noted as a player action and as so, will appear in the "/ledger search source:player" list.

### Observed/actual behavior

Nothing seems to be logged at all

### Steps/models to reproduce

Here's a quick video showcasing the process:
https://user-images.githubusercontent.com/71695040/145356866-4b406c07-bc6f-4423-bae1-ee9c82dd6f36.mp4

(i'm using carpet and lithium as you can see from my latest log. May be a compatibility issue)

### What operating system are you running

Linux

### Minecraft version

1.18 Server

### Ledger version

1.2.0

### Logs

https://github.com/QuiltServerTools/Ledger/files/7683304/latest.log

### Agreements

- [X] I am running the latest version of the mod.
- [X] My version of Minecraft is supported.
- [X] I have searched for and ensured there isn't already an open issue regarding this.

### Other

I'm using carpet and lithium as you can see from my latest log. So this may be a compatibility issue

### LINKED ISSUE #147: Log fire placed by lightning bolt
### Is your feature request related to a problem?

Yes, fire placed by lightning bolt does not log

### Describe the solution you'd like.

Fire placed by lightning bolt will be logged

### Describe alternatives you've considered.

None

### Agreements

- [X] I have searched for and ensured there isn't already an open issue regarding this.
- [X] I have ensured the feature I'm requesting isn't already in the latest supported mod version.

### Other

_No response_

### LINKED ISSUE #156: Log more actions, add more commands
### Is your feature request related to a problem?

I really like Ledger, but I feel it lhas very few logged actions, so I wanted to suggest more. I'm posting this because someone crafted an OP item (we unintentionally broke the recipe and made it basically free), but there's no way we can look up actions like dropped/picked up items, in order to facilitate our task.

Also, I would like to suggest some commands that could potentially increase efficiency when inspecting logs.

### Describe the solution you'd like.

The actions I think should be logged are:

- Explosions (TNT, Creepers, Ghasts, etc).
- Flint & steel and fire charge usage.
- Player interactions.
- Crops trampled by players.
- Items dropped or picked up by players.
- Placed/removed/modified paintings and item frames.
- Chat messages and commands used by players.
- Player sessions (logins/logouts).
- Player deaths.

The commands I think should be added are:

- /ledger near: Performs a lookup with a radius of 5 blocks.
- /ledger undo: Revert a rollback/restore via the opposite action.

### Describe alternatives you've considered.

Of all the actions that I consider should be logged, I only found one way to kinda check if a player had a certain item, and it was by creating a custom advancement, which is not ideal if you want to lookup other items, as you would have to create advancements for those too.

### Agreements

- [X] I have searched for and ensured there isn't already an open issue regarding this.
- [X] I have ensured the feature I'm requesting isn't already in the latest supported mod version.

### Other

_No response_

### LINKED ISSUE #168: Bed Destroy (Nether or End) marked as @gravity
### Expected behavior

The player who clicked the bed's name in the log.

### Observed/actual behavior

Upon inspection of the block, it is marked as:
@gravity broke White Bed

### Steps/models to reproduce

Place a bed in the End or Nether
Click on it to attempt to sleep in it, causing it to blow up via [Intentional Game Design]
Check the logs for the block
Observe the @gravity tag instead of the players name.

### What operating system are you running

Windows

### Minecraft version

1.18.2

### Ledger version

latest

### Logs

there is nothing in the log for you to see.

### Agreements

- [X] I am running the latest version of the mod.
- [X] My version of Minecraft is supported.
- [X] I have searched for and ensured there isn't already an open issue regarding this.

### Other

_No response_

### LINKED ISSUE #182: Ledger doesn't log blocks destroyed by Ender Dragon
### Expected behavior

Broken blocks are logged

### Observed/actual behavior

Broken blocks aren't logged

### Steps/models to reproduce

1) `/summon minecraft:ender_dragon`
2) `/fill ~-15 ~-15 ~-15 ~15 ~15 ~15 minecraft:stone`
3) Broken stone wasn't logged

Same with blocks that get destroyed by wither body while damage

### What operating system are you running

Windows

### Minecraft version

1.18.2

### Ledger version

1.2.4

### Logs

Nothing special in it

### Agreements

- [X] I am running the latest version of the mod.
- [X] My version of Minecraft is supported.
- [X] I have searched for and ensured there isn't already an open issue regarding this.

### Other

_No response_

### LINKED ISSUE #216: Track item frames and the item inside of them
### Is your feature request related to a problem?

Bases have item frames in their storage systems. I would like to be able to know that everything will be there once I rollback. I don't know if this is a bug, and oversight or just not possible, but currently item frames aren't tracked at all.
image

### Describe the solution you'd like.

It would be nice if ledger tracked placing and removing them, also placing and removing an item inside of them and the rotation of the item. I know this is a triple nested thing but all three scenarios are important.
Placing and removing the item frame for obvious reasons.
Placing and removing the item withing the frame because of storage systems
Rotating the item because there are types of combination locks that use item rotations in item frames to open and close.

### Describe alternatives you've considered.

I don't think there are any alternatives

### Agreements

- [X] I have searched for and ensured there isn't already an open issue regarding this.
- [X] I have ensured the feature I'm requesting isn't already in the latest supported mod version.

### Other

_No response_

### PR DESCRIPTION
**Pull request by nikijaz**
This PR expands logging capabilities and fixes some bugs. This also includes changes from #119 by Altirix.

**Added logging of:**
- Dyeing and shearing sheep (Including evoker as a source)
- Breaking a bed due to interaction (Fixes #168)
- Breaking a cactus, sugarcane, and chorus because of gravity
- Shooting a chorus flower
- Destroying a wither or (snow) golem statue
- Modifying a state of comparator or repeater
- Opening or closing a door, trapdoor, fence gate (Including redstone as a source)
- Breaking a lily pad with a boat
- Changing or destroying an armor stand (Fixes #129)
- Changing or destroying a frame (Fixes #216)
- Painting a dog or cat collar
- Breaking blocks by an ender dragon (Fixes #182)
- Dropping and picking up items (Fixes #156)
- Setting fire with lightning (Fixes #147)
- Making snow with a snow golem
- Turning a villager into a witch because of lightning
- Inserting or removing a book from a lectern

### CODE REVIEW

**Review by DrexHD — CHANGES_REQUESTED** [review]
Wow! This is a pretty big PR with a lot of changes.
I have added some small reviews to some parts of the code. I am putting some other things I would like to mention that didn't fit a specific piece of code here:

- `@ModifyArgs` is used in a couple of Mixins to get easy access to arguments. According to the Java Doc of ModifyArgs, it is pretty inefficient and this isn't really its intended use-case. I suggest using LocalCapture instead.
- This PR added a couple of mixins for plants breaking. Their source is `Gravity`, I think it would be great to add a separate source type for this, to allow better user blacklists (This would apply to CactusBlockMixin, SugarCaneBlockMixin, BambooBlockMixin, ChorusPlantBlock, and PointedDripstoneBlockMixin)
- This is probably a design choice, but since entity modify actions store entity positions, the entity is moved back to the action pos when rolling back (even if the position doesn't really matter, like dyeing).
- Item frame logging didn't work properly when I tried rolling back `@equip` actions
- When placing carved pumpkins (and probably wither skeleton skull) it first logs the block being broken by snowman spawning and then tries to log the placed block, but since it's no longer there it logs placing air (so the order is incorrect and the placed block is wrong)
- `ItemDropActionType` and `ItemPickUpActionType` should implement rollback / restore

Thanks for the PR!

**Comment by suerion** [comment]
I think on this Pull Request, there should also be the #202 projectile shoot event

File: src/main/kotlin/com/github/quiltservertools/ledger/database/Tables.kt [hunk 1]
```diff
@@ -65,6 +65,8 @@ object Tables {
         val sourcePlayer = optReference("player_id", Players.id)
         val extraData = text("extra_data").nullable()
         val rolledBack = bool("rolled_back").clientDefault { false }
+        val entityState = text("entity_state").nullable()
```
[the remark this answers is not in the document]
Comment (nikijaz) [reply src/main/kotlin/com/github/quiltservertools/ledger/database/Tables.kt]: Yeah, good idea, did it. But now there is a naming inconsistency, so maybe we should consider renaming it to `oldState` and `newState` in the future

**Review by DrexHD — CHANGES_REQUESTED** [review]
The new changes look pretty good 👍. I have left some more change request on some minor issues. 

I have also noticed an issue, when rolling back an armorstand that was first equipped and then killed. (The killing gets rolled back, but equipping doesn't, if both are rolled back in one operation)! I don't know what might cause this atm.
I think you *should* also remove the code duplication in `ItemDropActionType` and `ItemPickUpActionType` `rollback`/`restore`.

*Sorry for the slow responses, I was/am kind of busy currently*

**Comment by abitcher** [comment]
This isn't compatible with Goodall. There's an item called the Animal Crate that can capture animals, that isnt logged by this.

**Comment by abitcher** [comment]
This isn't compatible with Goodall. There's an item called the Animal Crate that can capture animals, that isnt logged by this.
image_2023-04-30_111441409

**Review by DrexHD — DISMISSED** [review]
Looks good to me!

**Comment by Dev0Louis** [comment]
Any news on this?

**Comment by Rozbiynk** [comment]
any updates?

**Review by DrexHD — DISMISSED** [review]

**Comment by DillonB07** [comment]
Would it be possible for this to be merged soon? Had an issue on my server that this would've solved but instead I had to go and move region files from an old backup as ledger hadn't logged the destruction.

**Comment by Rozbiynk** [comment]
i hope it will have 1.20.1 version

**Comment by TheDeathlyCow** [comment]
I've done some testing and can confirm that this branch works on 1.20.2 by just merging in [1.2.9](https://github.com/QuiltServerTools/Ledger/tree/1.2.9). 

There are, however, a couple of issues when merging in [update to 1.20.4](https://github.com/QuiltServerTools/Ledger/commit/a85a1e2d988d706f7dc625977f002ced969638a1). 

First, The `setBlockState` target for the injection [`TrapdoorBlockMixin#logTrapdoorInteraction`](https://github.com/nikijaz/Ledger/blob/more-logging/src/main/java/com/github/quiltservertools/ledger/mixin/blocks/TrapdoorBlockMixin.java) no longer exists. I think this can be fixed by changing the target to the `flip` call instead, as that seems to be the replacement. 

Second, the [`BedBlockMixin`](https://github.com/nikijaz/Ledger/blob/more-logging/src/main/java/com/github/quiltservertools/ledger/mixin/blocks/BedBlockMixin.java) has some invalid descriptors. It seems that `onBreak` now returns a `BlockState` so just changing the callback info param to `CallbackInfoReturnable<BlockState>` fixes that. 

Those are the only 2 errors I encountered with 1.20.4. Fixing them locally gets the mod running, and everything still seems to work.

**Review by PotatoPresident — CHANGES_REQUESTED** [review]

**Comment by PotatoPresident** [comment]
I can't help but feel that `sources` is being misused, but I don't really have a better alternative other than making a ton of different action types.

**Review by PotatoPresident — APPROVED** [review]

**Pull request merged**